framework/src/base/ComputeFullJacobianThread.C

```diff
@@ -148,6 +149,35 @@ ComputeFullJacobianThread::computeFaceJacobian(BoundaryID bnd_id)
     }
   }
 
+  /// done only when nonlocal integrated_bcs exist in the system
+  if (_fe_problem.checkNonlocalCouplingRequirement())
+  {
+    std::vector<std::pair<MooseVariable *, MooseVariable *> > & cne = _fe_problem.nonlocalCouplingEntries(_tid);
+    for (const auto & it : cne)
+    {
+      MooseVariable & ivariable = *(it.first);
+      MooseVariable & jvariable = *(it.second);
+
+      unsigned int ivar = ivariable.number();
+      unsigned int jvar = jvariable.number();
+
+      if (ivariable.activeOnSubdomain(_subdomain) && jvariable.activeOnSubdomain(_subdomain) && _integrated_bcs.hasActiveBoundaryObjects(bnd_id, _tid))
+      {
+        const std::vector<MooseSharedPointer<IntegratedBC> > & integrated_bcs = _integrated_bcs.getBoundaryObjects(bnd_id, _tid);
+        for (const auto & integrated_bc : integrated_bcs)
+        {
+          MooseSharedPointer<NonlocalIntegratedBC> nonlocal_integrated_bc = MooseSharedNamespace::dynamic_pointer_cast<NonlocalIntegratedBC>(integrated_bc);
```

Comment:
Lots of dynamic casting here. This might show up in profiling at some point. It would be better to store these objects in a separate containing in the warehouse. Let's not worry about it for now.